language/move-prover/spec-lang/src/env.rs

```diff
@@ -691,7 +694,41 @@ impl GlobalEnv {
 
     /// Adds a spec function to used_spec_funs set.
     pub fn add_used_spec_fun(&mut self, module_id: ModuleId, spec_fun_id: SpecFunId) {
-        self.used_spec_funs.insert(module_id.qualified(spec_fun_id));
+        let qid = module_id.qualified(spec_fun_id);
+        self.used_spec_funs.insert(qid);
+        self.propagate_move_fun_usage(qid);
+    }
+
+    /// Returns true if a spec fun is used in specs.
+    pub fn is_spec_fun_used(&self, module_id: ModuleId, spec_fun_id: SpecFunId) -> bool {
+        self.used_spec_funs
+            .contains(&module_id.qualified(spec_fun_id))
+    }
+
+    /// Adds an edge from the caller to the callee to the call graph.
+    pub fn add_edge_to_call_graph(
+        &mut self,
+        caller_mid: ModuleId,
+        caller_fid: SpecFunId,
+        callee_mid: ModuleId,
+        callee_fid: SpecFunId,
+    ) {
+        self.call_graph
+            .entry(caller_mid.qualified(caller_fid))
+            .or_insert_with(BTreeSet::new)
+            .insert(callee_mid.qualified(callee_fid));
+    }
+
+    /// Runs DFS to propagate the usage of move functions from callers
+    /// to callees on the call graph.
+    pub fn propagate_move_fun_usage(&mut self, qid: QualifiedId<SpecFunId>) {
+        if let Some(neighbors) = self.call_graph.get(&qid) {
+            neighbors.clone().iter().for_each(|n| {
+                if self.used_spec_funs.insert(*n) {
```

Comment:
Add a comment here that via the if here, you deal with recursion.

Reply:
Also, you only do the callgraph for move functions, right? Should then lokely call the field move_fun_call_graph.

Reply:
This call graph also includes all spec functions(both actual spec fun and move-fun-converted spec fun) although we currently don't use the information of whether an actual spec fun is used in later translation. I will change the function name or the field name to be more consistent with each other.